README.md
# ember-i18next
![Build Status](https://github.com/OCTRI/ember-i18next/workflows/CI/badge.svg)

## About

An [Ember CLI](http://www.ember-cli.com/) addon for internationalizing Ember.js applications using the [i18next](http://i18next.com/) library. The addon provides an Ember service that wraps i18next and a Handlebars helper for displaying localized text in templates.

## Compatibility

ember-i18next supports current Ember (release, beta, canary) and the last two LTS releases.

## Installation

To install with Ember CLI:

```bash
ember install ember-i18next
```

## Configuration

### Configuring i18next

To configure the [i18next options](https://www.i18next.com/configuration-options.html) and the [XHR backend options](https://github.com/i18next/i18next-xhr-backend#backend-options), add them to your `environment.js`:

```javascript
// ...
let ENV = {
  // ...
  i18nextOptions: {
    // any options supported by i18next
    backend: {
      // any options supported by i18next-xhr-backend
    },
    // set to true if you want to catch errors when initializing i18next
    rejectError: false
  },
  APP: {
    // ...
  }
}
```

If you do not specify any options, the default i18next options will be used.

### Initializing i18next

To initialize the i18next library, call the i18n service's `initLibraryAsync` method. This method returns a promise that resolves when the library finishes initializing, so if you call it in one of the model hook methods (`beforeModel`, `model`, `afterModel`), the application will enter the loading substate until i18next is initialized. The application route may be a good place to do this.

#### Handling Initialization Errors

By default, this addon will catch any initialization errors and display a warning in the console. If you want to catch and handle the errors, set the `rejectError` option to true and handle the errors in a `catch` block.

```javascript
// app/routes/application.js
import Route from '@ember/routing/route';

export default Route.extend({
  
  beforeModel() {
    return this.get('i18n')
      .initLibraryAsync()
      // only if `rejectError` option is set to true
      .catch(errors => this.transitionTo('my-error-page', errors));
  }

});
```

### Including Locale Files

By default, i18next loads locale files from the server asynchronously from the server path configured using the XHR backend's [`loadPath`](https://github.com/i18next/i18next-xhr-backend#backend-options) configuration option. To copy your application's locale resources from your source tree to the expected path during build, modify the application's `ember-cli-build.js`:

```javascript
const EmberApp = require('ember-cli/lib/broccoli/ember-app');
const funnel = require('broccoli-funnel');

module.exports = function(defaults) {
  const app = new EmberApp(defaults, {
    // build options
  });

  const locales = funnel('app/locales', {
    srcDir: '/',
    destDir: '/locales'
  });

  // other configuration, app.import() calls, etc. ...

  return app.toTree(locales);
}
```

In this example, [Broccoli Funnel](https://github.com/broccolijs/broccoli-funnel) recursively copies locale files from the application's `app/locales/` directory to `dist/locales/` when the application is built.

## Use

### Service

If you need to produce translated strings in routes, components or controllers, you can [inject the `i18n` service](https://guides.emberjs.com/v3.17.0/applications/dependency-injection/#toc_ad-hoc-injections). This will then give access to [i18next's `t()` function](https://www.i18next.com/essentials.html) in your code. For example:

```javascript
// app/components/example-component.js
import Component from '@ember/component';
import { computed } from '@ember/object';
import { inject as service } from '@ember/service';

export default Component.extend({
  i18n: service(),
  messages: someObject,

  messageCount: computed('messages', function () {
    const i18n = this.get('i18n');
    const count = this.get('messages.count');
    return i18n.t('messages.count', { msgCount: count });
  }
});
```

For convenience, a mixin is provided that injects the i18n service and adds a `t()` function to the including class. For example, the above could be accomplished using the mixin like this:

```javascript
// app/components/example-component.js
import Component from '@ember/component';
import { computed } from '@ember/object';

import I18nMixin from 'ember-i18next/mixins/i18n';

export default Component.extend(I18nMixin, {
  messages: someObject,

  messageCount: computed('messages', function () {
    const count = this.get('messages.count');
    return this.t('messages.count', { msgCount: count })
  });
});
```

### Handlebars Helper

You can access your app's translations in templates using the `t` helper:

```handlebars
<button type="button">{{t 'button.text'}}</button>
```

Pass values to be interpolated into the translation as [hash arguments](http://handlebarsjs.com/guide/expressions.html#helpers-with-hash-arguments). For example, for a translation that includes an interpolated `{{count}}` value:

```handlebars
<div>{{t 'messages.count' count=model.messageCount}}</div>
```

### Macro

You can create computed properties that watch for locale changes:

```js
import { translationMacro as t } from "ember-i18next";

export default Ember.Component.extend({
  // A simple translation.
  title: t('user.edit.title'),

  followersCount: 1,
  count: Ember.computed.alias('followersCount'),

  // A translation with interpolations. This computed property
  // depends on `count` and will send `{ count: this.get('count') }`
  // in to the translation.
  followersTitle: t('user.followers.title', 'count')
});
```

The macro relies on this.get('i18n') being the `service:i18n`. Make sure it is injected somehow.

### Current Locale

The current locale is exposed via i18n service's `locale` property. To change the language that the application is displayed in, simply set this property, and all of the text displayed using the `t` helper will be updated. For example, triggering the following controller action would update all of the text to Thai:

```javascript
// app/controllers/example-controller.js
import Controller from '@ember/controller';
import I18nMixin from 'ember-i18next/mixins/i18n';

export default Controller.extend(I18nMixin, {
  actions: {
    showThai() {
      this.set('i18n.locale', 'th-TH');
    }
  }
});
```

### Pre- and Post-Init Actions

Changing the locale causes i18next to be reinitialized, which can destroy state. For example, if you have used `addResources` to load additional localization strings, they will be lost during initialization. To handle management of state around library initialization, you can register actions to perform before and after library initialization with the i18n service using the `registerPreInitAction` and `registerPostInitAction` methods.

```javascript
import Route from '@ember/routing/route';
import I18nMixin from 'ember-i18next/mixins/i18n';

export default Route.extend(I18nMixin, {
  init() {
    this._super();
    const i18n = this.get('i18n');

    // my-fancy-action is a name that can be used to unregister the action later
    i18n.registerPostInitAction('my-fancy-action', oldLang => {
      // do preloading
    });
  }
}
```

Pre- or post-init actions may return a promise. If any of the actions returns a promise, the service will wait for the promises to resolve before moving on. If pre-init actions return promises, the service will wait for them to resolve before initializing i18next. If post-init actions return promises, the service will wait for them to resolve before notifying the application about the change in locale.

The new locale is passed as a parameter to pre-init actions and the old locale to post-init actions. These parameters are `undefined` when the the i18n service's `initLibraryAsync` method is called.

Finally, actions may be unregistered using the `unregisterPreInitAction` and `unregisterPostInitAction` methods. To unregister the post-init action from the previous example, you would do the following:

```javascript
// ...
i18n.unregisterPostInitAction('my-fancy-action');
// ...
```

## Testing

### Acceptance Tests

No special configuration is required for acceptance testing, although it may be convenient to configure the default locale and preload locales for use in the tests. For example:

```javascript
// in environment.js
if (environment === 'test') {
  // ...
  ENV.i18nextOptions.lng = 'en-US';
  ENV.i18nextOptions.preload = ['en-US', 'th-TH'];
}
```

### Component Integration Tests

If you need to make assertions about the text rendered in component integration tests, you can [initialize the i18n service in a `beforeEach` hook](https://github.com/OCTRI/ember-i18next/issues/31).

```javascript
moduleForComponent('some-component', 'Integration | Component | some component', {
  integration: true,

  beforeEach(assert) {
    const done = assert.async();
    this.inject.service('i18n');
    this.get('i18n').initLibraryAsync().then(done);
  }
});
```

### Unit Tests

Unit tests for objects that inject the i18n service or use the `{{t}}` helper should add them to the `needs` array.

```javascript
moduleForComponent('other-component', 'Unit | Component | other component', {
  unit: true,
  needs: ['service:i18n', 'helper:t'],
});
```

## Collaborating

Contributions are happily accepted. Make sure that your pull request includes tests and your JavaScript source is styled as described in the [Ember.js JavaScript style guide](https://github.com/emberjs/ember.js/blob/master/STYLEGUIDE.md).

## Acknowledgements

Early versions of this addon were strongly influenced by the  [ember-cli-i18n](https://github.com/dockyard/ember-cli-i18n) addon.


Changelog.md
# Changelog

## 5.0.0

Updates dependencies and adds the ability to catch errors thrown during initialization.

### Breaking

ember-i18next 5.0.0 updates the i18next dependency to version 19, which contains breaking changes. Review [the i18next migration guide](https://www.i18next.com/misc/migration-guide.html) prior to upgrading.

### New Features

This release adds the ability to customize the behavior when a promise rejects in `initLibraryAsync`. Previously ember-i18next would only log the error to the console. For compatibility with previous versions, this is still the default behavior. However, you can now handle errors in a catch block if the `rejectError` configuration property is truthy. See the README for details.

Thanks to @mpirio for implementing this.

### Changes Included

* PR #100: Documentation updates.
* PR #79, #98: Upgrade to latest i18next and i18next-xhr-backend.
* PR #95: Add configuration property to allow handling errors raised by `initLibraryAsync`. Thanks to @mpirio for this PR.
* PR #88: Continuous integration changes.
* Many dependency updates.


## 4.0.0

Updates i18next and i18next-xhr-backend. No changes to ember-i18net API.

### Breaking

ember-i18next 4.0.0 contains no changes to the ember-i18next API, but it updates the i18next dependency to version 15, which contains breaking changes. Review [the i18next migration guide](https://www.i18next.com/misc/migration-guide.html) prior to upgrading.

### Changes Included

* PR #62: Upgrade to latest i18next and i18next-xhr-backend. Thanks to @mpirio for this PR.
* PR #63: Use promises returned by asynchronous i18next method calls.

## 3.0.2

Updates dependencies.

### Changes Included

* PR #61: Update to Ember 3.7.

## 3.0.1

Updates dependencies.

### Changes Included

* PR #60: Update merge to fix prototype pollution vulnerability.

## 3.0.0

Updates dependencies and migrates to use the new Ember testing APIs.

### Breaking

ember-i18next 3.0.0 drops support for Ember 1.13. The addon has been tested with Ember 2.16 LTS, 2.18 LTS and Ember 3.x, but it should work down to version 2.3.

Version 3.0.0 also updates the i18next dependency to version 11, which contains breaking changes. Review [the i18next migration guide](https://www.i18next.com/misc/migration-guide.html) prior to upgrading.

### Changes Included

* PR #56: Use the new Ember testing API.
* PR #57: Update addon dependency to Ember 3.4.
* PR #58: Update to the latest i18next and i18next-xhr-backend.

## 2.1.1

Updates dependencies.

### Changes Included

* PR #55: Update addon dependencies.

## 2.1.0 2018/04/24

Updates depdendencies and adds a new computed property macro for translated text. See README.md for details on using the new macro.

### Changes Included

* PR #54: Update addon depdendency to Ember 3.0.
* PR #52: Add computed property macro.

## 2.0.0 2018/03/01

Updates dependencies and uses the [Ember.js modules API introduced in Ember 2.16](https://emberjs.com/blog/2017/10/11/ember-2-16-released.html).

### Breaking

ember-i18next 2.0.0 updates the i18next dependency to version 10, which contains breaking changes. Review the README.md and [the i18next migration guide](https://www.i18next.com/misc/migration-guide.html) prior to upgrading.

### Changes included

* PR #45: Remove use of Bower.
* PR #47: Remove use of Browserify.
* PR #48: Move code to the `addon` directory.
* PR #50: Update dependencies and use the new Ember modules API.

## 2.0.0-beta.2 2016/09/05

### Breaking

Ember-i18next 2 updates the i18next dependency to version 3, which contains breaking changes. Review the README.md and [the i18next migration guide](http://i18next.com/docs/migration/) prior to upgrading.

### Changes included

* PR #41: For compatibility with Ember 2.9, do not modify the positional parameters passed to the `{{t}}` helper.

## 2.0.0-beta.1 2016/07/07

### Breaking

Updates the i18next dependency to version 3, which contains breaking changes. Review the README.md and [the i18next migration guide](http://i18next.com/docs/migration/) prior to upgrading.

### Changes included

* PR #32: Update to use i18next 3. Thanks to @john-coffey for implementing this.
* PR #33: Return a safe string from the `{{t}}` helper, allowing translations to contain markup.
* Miscellaneous code and documentation cleanup: #28, #29, #31, #36.

## 1.1.1 2016/01/15

Specify the version string to add to the bower.json when the addon is installed, preventing broken builds (see issue #25).

## 1.1.0 2015/12/16

Now passes a locale to the pre- and post-init actions, allowing them to perform work that depends on the locale, such as loading or unloading additional resources. The new locale is passed to pre-init actions, and the old locale is passed to post-init actions. See #24 for details. Thanks to @mpirio for making this happen.

## 1.0.0 2015/08/27

### Breaking

Includes changes needed to work with Ember 1.13, 2.0, and 2.1, and canary. Does not support versions of Ember prior to 1.13.

## 1.0.0-beta.1 Not released.

Included changes needed to work with Ember 1.13 and Ember 2.0 and fixes deprecations.

## 0.3.1 2015/06/09

Includes updates due to transfer to OCTRI organization. No code changes are included.

## 0.3.0 2015/06/09

Adds hooks that run around initialization of the i18next library. Any registered actions will be run when the i18next library is initialized as a result of a call to `initLibraryAsyc` or changing the locale.

To register a function to run before the i18next library is initialized, use the i18n service's `registerPreInitAction` method. Similarly, to register a function to run after the i18next library is initialized, use the `registerPostInitAction` method. Pre- and post-init actions can be unregistered using the `unregisterPreInitAction` / `unregisterPostInitAction` methods.

See the [README](https://github.com/heathharrelson/ember-i18next/blob/33e55a857b4daadfc25bbbe8322053c62f731425/README.md) for details and examples.

## 0.3.0-beta.3 2015/05/27


config/ember-try.js
'use strict';

const getChannelURL = require('ember-source-channel-url');

module.exports = function() {
  return Promise.all([
    getChannelURL('release'),
    getChannelURL('beta'),
    getChannelURL('canary'),
  ]).then((urls) => {
    return {
      useYarn: true,
      scenarios: [
        {
          name: 'ember-1.13',
          bower: {
            dependencies: {
              'ember': '1.13.13'
            }
          },
          npm: {
            devDependencies: {
              'ember-source': null
            }
          }
        },
        {
          name: 'ember-lts-2.12',
          npm: {
            devDependencies: {
              'ember-source': '~2.12.0'
            }
          }
        },
        {
          name: 'ember-lts-2.16',
          npm: {
            devDependencies: {
              'ember-source': '~2.16.0'
            }
          }
        },
        {
          name: 'ember-lts-2.18',
          npm: {
            devDependencies: {
              'ember-source': '~2.18.0'
            }
          }
        },
        {
          name: 'ember-release',
          npm: {
            devDependencies: {
              'ember-source': urls[0]
            }
          }
        },
        {
          name: 'ember-beta',
          npm: {
            devDependencies: {
              'ember-source': urls[1]
            }
          }
        },
        {
          name: 'ember-canary',
          npm: {
            devDependencies: {
              'ember-source': urls[2]
            }
          }
        },
        {
          name: 'ember-default',
          npm: {
            devDependencies: {}
          }
        }
      ]
    };
  });
};
